### PR TITLE
More dv queries supported for file/link suggestions

### PR DESCRIPTION
**Pull request by Krakor92**
I started this PR as a result of a conversation on the Obsidian Discord (https://discord.com/channels/686053708261228577/1014259487445622855/1067595344671682621).

First I tried to write the dvjs query to achieve it. Here it is

````
```dataviewjs
[...dv.pages().map(p => p.subject).distinct()].reduce((acc, cur) => { if (cur) {return [...acc, cur]} return acc }, [])
```
````

It was messy and i kept having errors. So I took a look in the code and noticed that the `getFiles` method of `FileField.ts` and `MultiFileField.ts` expected the `getResults()` function to return an object with a `values` field.

I still don't understand what the proxy target returned by `dv.pages()` contains because at first glance, the `.values` property looks like a perfectly normal array but I can't call `.filter()` on it without triggering an error for some reason :/. I can't call `instanceof Proxy` to differentiate it from a normal array but I found another way to find out: If `results.values` is an object then `results` is a proxy, otherwise it is a regular array (I'm doing a `reduce` to remove the `undefined` results in that case). 

On top of that, I simply put additional `if` checks inside the `map` callback so that if the dataview query written by the user returns an array of `TFile`, then it will work (as before) but if they want to call the `.map` method next to the `dv.pages()` to return an array of `Link` instead, that will also work.

To summarise: Before this pr, the only thing you could write inside the optional dv query of a file/multiFile field was a single `dv.pages()` with a predicate inside. So you couldn't implement custom filtering that fetches its values on the fly. Now the following should also work:

````
```dataviewjs
// For MultiFile field
[...dv.pages().map(p => p.subject)].flat()

// For SingleFile field
dv.pages().map(p => p.subject)
```
````

Also, I couldn't help but isolate this piece of shared logic into its own file, then import it into both `FileField.ts` and `MultiFileField.ts` ^^'

I'd be happy to hear your comments on this.

### CODE REVIEW

**Comment by mdelobelle** [comment]
Hi, that's a great improvement. could you add some user documentation for it here: https://github.com/mdelobelle/metadatamenu/blob/master/docs/settings.md#file-multifile-options ?

**Comment by Krakor92** [comment]
To reduce friction on the end-user side, I rewrote the getFiles function so that it can handle tricky situations (instead of assuming that the end user will figure it out on his own...)

For the following example, I assume that the end user wants to force the suggester to show him only links he has already defined on a specific field in his vault.

Let's assume that what's below is what is set as dv query in the field `manga`'s settings
```
dv.pages().map(p => p.manga)
```

So, in my previous commit, if you had, in a file, a property containing an array of links like that `manga:: [[One Piece]], [[Naruto]]`, nothing appeared in the suggestion because Arrays weren't supported by the getFiles function. So the end user had to modify its dv query as follows to make it work: 

```
[...dv.pages().map(p => p.manga)].flat()
```

For a non-developer, this introduces some error-prone weird syntax, so I think it's better to have it handled directly in the codebase. And that's what I did. Now the first query works as expected for the end user.


---

Looking through the code, I also noticed how similar `FileField.ts` and `MutiFileField.ts` are. About 95% of their content is identical. At first I didn't really understand the point of separating these two field types, but I kind of get it now. Although it would be much cleaner to abstract most of their code inside a common file, what do you think?

+ Also I've fixed the sorting part example in the documentation

**Comment by mdelobelle** [comment]
Thanks a lot, i'll merge that and release in 0.4.18

> Looking through the code, I also noticed how similar FileField.ts and MutiFileField.ts are. About 95% of their content is identical. At first I didn't really understand the point of separating these two field types, but I kind of get it now. Although it would be much cleaner to abstract most of their code inside a common file, what do you think?

You're absolutely right, it was pure lazyness 🙂 . I've refactored the common code in AbstractFileBasedField and included your FileFieldUtils in it.
